arangod/Aql/Ast.cpp

```diff
@@ -2624,170 +2624,218 @@ size_t Ast::countReferences(AstNode const* node, Variable const* search) {
   return result.count;
 }
 
-/// @brief determines the top-level attributes referenced in an expression,
-/// grouped by variable name
-TopLevelAttributes Ast::getReferencedAttributes(AstNode const* node,
-                                                bool& isSafeForOptimization) {
-  TopLevelAttributes result;
-
+// @brief determines the attributes (and their subattributes) referenced in an
+// expression for the specified out variable and a subattribute from it
+// (e.g. p.edges[0].abc, p.vertices[*].abc). if the subattribute is empty,
+// then it will be ignored.
+// we need to take care of the following example cases here:
+// - v                  => no projections
+// - v[0]               => no projections
+// - v[0][1]            => no projections
+// - v.x                => ["x"]
+// - v[0].x             => no projections
+// - v[0][1].x          => no projections
+// - v[0].x[1].x        => no projections
+// - v.x.y              => [["x", "y"]]
+// - p                  => no projections
+// - p[0].vertices      => no projections
+// - p.vertices         => no projections
+// - p.vertices[0][1].x => no projections
+// - p.vertices.x       => no projections
+// - p.vertices[0].x    => ["x"]
+// - p.vertices[0].x.y  => [["x", "y"]]
+bool Ast::getReferencedAttributesRecursive(
+    AstNode const* node, Variable const* variable,
+    std::string_view expectedAttribute,
+    std::unordered_set<arangodb::aql::AttributeNamePath>& attributes) {
   // traversal state
-  char const* attributeName = nullptr;
-  size_t nameLength = 0;
-  isSafeForOptimization = true;
-
-  auto visitor = [&](AstNode const* node) {
-    if (node->type == NODE_TYPE_ATTRIBUTE_ACCESS) {
-      attributeName = node->getStringValue();
-      nameLength = node->getStringLength();
-      return true;
-    }
+  struct TraversalState {
+    Variable const* variable;
+    bool isSafeForOptimization;
+    std::string_view expectedAttribute;
+    containers::SmallVector<AstNode const*, 128> seen;
+    std::unordered_set<arangodb::aql::AttributeNamePath>& attributes;
+  };
 
-    if (node->type == NODE_TYPE_REFERENCE) {
-      // reference to a variable
-      if (attributeName == nullptr) {
-        // we haven't seen an attribute access directly before...
-        // this may have been an access to an indexed property, e.g value[0] or
-        // a reference to the complete value, e.g. FUNC(value)
-        // note that this is unsafe to optimize this away
-        isSafeForOptimization = false;
-        return true;
+  TraversalState state{variable,
+                       /*isSafeForOptimization*/ true,
+                       expectedAttribute,
+                       {},
+                       attributes};
+
+  auto evalStack = [&state]() -> bool {
+    TRI_ASSERT(!state.seen.empty());
+    AstNode const* top = state.seen.back();
+    TRI_ASSERT(top->type == NODE_TYPE_REFERENCE);
```

Comment:
Agree on this one, the below method does not protect us from illegally handed in Node in the first place and could certainly pass through a different type (e.g. ATTRIBUTE_ACCESS on top level) This would indicate a broken optimizer rule, calling this with invalid values, or a wrongly parsed query.